[TACHYON-1132] Use options for Tachyon object creation

jsimsa opened the pull request:
This PR addresses: https://tachyon.atlassian.net/browse/TACHYON-1132?filter=-2

jsimsa commented:
@calvinjia @apc999: missing some JavaDoc and unit tests, but I am heading out now :-)

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6011/
Test PASSed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6018/<h2>Build result: FAILURE</span></h2>[...truncated 2168 lines...][JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/swift/pom.xml to org.tachyonproject/tachyon-underfs-swift/0.8.0-SNAPSHOT/tachyon-underfs-swift-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/swift/target/tachyon-underfs-swift-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-swift/0.8.0-SNAPSHOT/tachyon-underfs-swift-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/swift/target/tachyon-underfs-swift-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-swift/0.8.0-SNAPSHOT/tachyon-underfs-swift-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/swift/target/tachyon-underfs-swift-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-swift/0.8.0-SNAPSHOT/tachyon-underfs-swift-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/examples/pom.xml to org.tachyonproject/tachyon-examples/0.8.0-SNAPSHOT/tachyon-examples-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/s3/pom.xml to org.tachyonproject/tachyon-underfs-s3/0.8.0-SNAPSHOT/tachyon-underfs-s3-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/s3/target/tachyon-underfs-s3-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-s3/0.8.0-SNAPSHOT/tachyon-underfs-s3-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/s3/target/tachyon-underfs-s3-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-s3/0.8.0-SNAPSHOT/tachyon-underfs-s3-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/s3/target/tachyon-underfs-s3-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-s3/0.8.0-SNAPSHOT/tachyon-underfs-s3-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/pom.xml to org.tachyonproject/tachyon-clients/0.8.0-SNAPSHOT/tachyon-clients-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/unshaded/pom.xml to org.tachyonproject/tachyon-client-unshaded/0.8.0-SNAPSHOT/tachyon-client-unshaded-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/servers/pom.xml to org.tachyonproject/tachyon-servers/0.8.0-SNAPSHOT/tachyon-servers-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/pom.xml to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/hdfs/target/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-hdfs/0.8.0-SNAPSHOT/tachyon-underfs-hdfs-0.8.0-SNAPSHOT-javadoc.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/pom.xml to org.tachyonproject/tachyon-parent/0.8.0-SNAPSHOT/tachyon-parent-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/clients/client/pom.xml to org.tachyonproject/tachyon-client/0.8.0-SNAPSHOT/tachyon-client-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/shell/pom.xml to org.tachyonproject/tachyon-shell/0.8.0-SNAPSHOT/tachyon-shell-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/pom.xml to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Tachyon-Pull-Request-Builder/underfs/glusterfs/target/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jar to org.tachyonproject/tachyon-underfs-glusterfs/0.8.0-SNAPSHOT/tachyon-underfs-glusterfs-0.8.0-SNAPSHOT-javadoc.jarchannel stopped
Test FAILed.

jsimsa commented:
I added unit tests and JavaDoc. There are some TODOs and I would prefer to finish this up myself on Sunday evening if that's okay with you.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6022/<h2>Failed Tests: <span class='status-failure'>1</span></h2><h3><a name='org.tachyonproject:tachyon-integration-tests' /><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6022/org.tachyonproject$tachyon-integration-tests/testReport'>org.tachyonproject:tachyon-integration-tests</a>: <span class='status-failure'>1</span></h3><ul><li><a href='https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6022/org.tachyonproject$tachyon-integration-tests/testReport/tachyon.master.file/FileSystemMasterIntegrationTest/lastModificationTimeCreateFileTest/'><strong>tachyon.master.file.FileSystemMasterIntegrationTest.lastModificationTimeCreateFileTest</strong></a></li></ul>
Test FAILed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6048/
Test PASSed.

jsimsa commented:
This PR is now ready for a review.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6050/
Test PASSed.

calvinjia commented:
@jsimsa why do we move the `BlockMasterClient` from common?

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/TachyonFS.java:
```diff
@@ -297,19 +299,24 @@ public synchronized int createDependency(List<String> parents, List<String> chil
+      if (blockSizeBytes > 0) {
+        UnderStorageType underStorageType =
+            mTachyonConf.getEnum(Constants.USER_FILE_WRITE_TYPE_DEFAULT, WriteType.class)
+                .isThrough() ? UnderStorageType.SYNC_PERSIST : UnderStorageType.NO_PERSIST;
```
Should `writeType` be used to determine `TachyonStorageType`?

jsimsa replied:
I figured if someone is still using the deprecated TachyonFS API, they might still be using the deprecated way of specifying whether to persist the data or not. Happy to change this if you think that makes more sense.

calvinjia replied:
Oh I meant if we should also set `TachyonStorageType` if they are using the old API?

yupeng9 replied:
NIT: should this be extracted into a function, since it repeats below?

jsimsa replied:
@calvinjia currently, neither `create` nor `mkdir` take `TachyonStorageType` as an option. The `getInStream` and `getOutStream` methods in the old `TachyonFile` API do.

jsimsa replied:
@yupeng9 I would prefer to inline this logic as it is only one statement and inlining it improves readability.

jsimsa commented:
The FileSystemMasterClient now imports ClientOptions, which are a client-side concept. So I moved the FileSystemMasterClient to the client module to avoid introducing a dependency from common to clients. For consistency, I moved the other clients too. Is there a reason not to do that?

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/file/TachyonFileSystem.java:
```diff
@@ -157,8 +157,10 @@ public FileOutStream getOutStream(TachyonURI path)
-    CreateOptions createOptions = (new CreateOptions.Builder(ClientContext.getConf()))
-        .setBlockSize(options.getBlockSize()).setRecursive(true).setTTL(options.getTTL()).build();
+    CreateOptions createOptions =
+        (new CreateOptions.Builder(ClientContext.getConf()))
```
Are the outer parens required?

jsimsa replied:
Fixed.

calvinjia commented:
@jsimsa Hmm I see, currently it will be okay since servers depends on clients, but ideally they would be separate

calvinjia commented on clients/unshaded/src/main/java/tachyon/client/file/options/MkdirOptions.java:
```diff
@@ -44,6 +50,15 @@ public Builder setRecursive(boolean recursive) {
+    public Builder setUnderStorageType(UnderStorageType underStorageType) {
+      mUnderStorageType = underStorageType;
+      return this;
+    }
```
Should we call this `setPersistence(boolean)` to make it more straightforward?

jsimsa replied:
How would you represent the ASYNC_PERSIST type in that case?

calvinjia replied:
It seems like we end up converting that to `false`? (`options.setPersisted(mUnderStorageType.isSyncPersist())`)

jsimsa replied:
That is true, but that is simply what the current implementation does. Using UnderStorageType for the argument type (as opposed to boolean) allows us to implement support for ASYNC_PERSIST in the future, without changing the CreateOptions / MkdirOptions / OutStreamOptions API.

calvinjia replied:
Okay that makes sense, I originally thought there were only two outcomes we would consider.

yupeng9 commented on clients/unshaded/src/main/java/tachyon/client/file/FileOutStream.java:
```diff
@@ -90,10 +90,6 @@ public FileOutStream(long fileId, OutStreamOptions options) throws IOException {
       String fileName = PathUtils.temporaryFileName(fileId, mNonce, mUfsPath);
       UnderFileSystem ufs = UnderFileSystem.get(fileName, ClientContext.getConf());
-      String parentPath = (new TachyonURI(mUfsPath)).getParent().getPath();
-      if (!ufs.exists(parentPath) && !ufs.mkdirs(parentPath, true)) {
```
Why is this check no longer needed?

jsimsa replied:
Because as of this PR, the path is guaranteed to be created on the master-side when the underlying file is created.

yupeng9 commented:
Thanks for the improvement! It overall looks good to me.

jsimsa commented:
Thanks for the review guys. I addressed your comments. PTAL

yupeng9 commented:
LGTM

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/6059/
Test PASSed.

calvinjia commented:
LGTM
